File: teacher/app/src/main/java/com/instructure/teacher/services/FileUploadService.java
/*
 * Copyright (C) 2017 - present  Instructure, Inc.
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, version 3 of the License.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.instructure.teacher.services;

import android.accounts.NetworkErrorException;
import android.app.IntentService;
import android.app.NotificationManager;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.NotificationCompat;

import com.instructure.canvasapi2.managers.FileFolderManager;
import com.instructure.canvasapi2.managers.UserManager;
import com.instructure.canvasapi2.models.FileUploadParams;
import com.instructure.canvasapi2.models.RemoteFile;
import com.instructure.pandautils.models.FileSubmitObject;
import com.instructure.pandautils.utils.Const;
import com.instructure.teacher.R;
import com.instructure.teacher.utils.FileUploadUtils;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Locale;

public class FileUploadService extends IntentService {

    public FileUploadService() {
        this(FileUploadService.class.getSimpleName());
    }

    public FileUploadService(final String name) {
        super(name);
    }

    private int mUploadCount;
    private static final int NOTIFICATION_ID = 1;
    private boolean isCanceled = false;

    // Upload broadcasts
    public static final String ALL_UPLOADS_COMPLETED = "ALL_UPLOADS_COMPLETED";
    public static final String UPLOAD_COMPLETED = "UPLOAD_COMPLETED";
    public static final String UPLOAD_ERROR = "UPLOAD_ERROR";

    // Upload Actions
    public static final String ACTION_ASSIGNMENT_SUBMISSION = "ACTION_ASSIGNMENT_SUBMISSION";
    public static final String ACTION_MESSAGE_ATTACHMENTS = "ACTION_MESSAGE_ATTACHMENTS";
    public static final String ACTION_COURSE_FILE = "ACTION_COURSE_FILE";
    public static final String ACTION_USER_FILE = "ACTION_USER_FILE";
    public static final String ACTION_DISCUSSION_ATTACHMENT = "ACTION_DISCUSSION_ATTACHMENT";
    public static final String ACTION_CANCEL_UPLOAD = "ACTION_CANCEL_UPLOAD";
    public static final String ACTION_SG_MEDIA_COMMENT = "actionSgMediaComment";

    public static final String MESSAGE_ATTACHMENT_PATH = "conversation attachments";
    public static final String DISCUSSION_ATTACHMENT_PATH = "discussion attachments";

    private NotificationCompat.Builder mNotificationBuilder;
    private NotificationManager mNotificationManager;

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        // Handle 'cancel' action in onStartCommand instead of onHandleIntent, because threading.
        if (ACTION_CANCEL_UPLOAD.equals(intent.getAction())) isCanceled = true;
        return super.onStartCommand(intent, flags, startId);
    }

    @Override
    protected void onHandleIntent(Intent intent) {

        // Skip if canceled
        if (isCanceled) return;

        final String action = intent.getAction();
        final Bundle bundle = intent.getExtras();
        final ArrayList<FileSubmitObject> fileSubmitObjects = bundle.getParcelableArrayList(Const.FILES);

        mUploadCount = fileSubmitObjects == null ? 0 : fileSubmitObjects.size();
        showNotification(mUploadCount);

        startUploads(action, fileSubmitObjects, bundle);
    }

    //region Upload functionality

    private void startUploads(String action, ArrayList<FileSubmitObject> fileSubmitObjects, Bundle bundle) {
        ArrayList<RemoteFile> attachments = new ArrayList<>();

        final long courseId = bundle.getLong(Const.COURSE_ID);
        final long assignmentId = bundle.getLong(Const.ASSIGNMENT_ID);
        final long conversationId = bundle.getLong(Const.CONVERSATION);

        Long parentFolderId = null;
        if (bundle.containsKey(Const.PARENT_FOLDER_ID)) {
            parentFolderId = bundle.getLong(Const.PARENT_FOLDER_ID);
        }

        try {
            for (int i = 0; i < fileSubmitObjects.size(); i++) {

                // Handle upload cancellation
                if (isCanceled) {
                    stopForeground(true);
                    mNotificationManager.cancel(NOTIFICATION_ID);
                    stopSelf();
                    return;
                }

                RemoteFile attachment = null;
                FileSubmitObject fso = fileSubmitObjects.get(i);

                updateNotificationCount(fso.getName(), i + 1);

                if (ACTION_ASSIGNMENT_SUBMISSION.equals(action)) {
                    attachment = uploadSubmissionFiles(courseId, assignmentId, fso.getSize(), fso.getName(), fso.getContentType(), fso.getFullPath());
                } else if (ACTION_COURSE_FILE.equals(action)) {
                    attachment = uploadCourseFiles(courseId, parentFolderId, fso.getSize(), fso.getName(), fso.getContentType(), fso.getFullPath(), null);
                } else if (ACTION_USER_FILE.equals(action)) {
                    attachment = uploadUserFiles(fso.getSize(), fso.getName(), fso.getContentType(), fso.getFullPath(), null, parentFolderId);
                } else if (ACTION_MESSAGE_ATTACHMENTS.equals(action)) {
                    attachment = uploadUserFiles(fso.getSize(), fso.getName(), fso.getContentType(), fso.getFullPath(), MESSAGE_ATTACHMENT_PATH, null);
                } else if (ACTION_DISCUSSION_ATTACHMENT.equals(action)) {
                    attachment = uploadUserFiles(fso.getSize(), fso.getName(), fso.getContentType(), fso.getFullPath(), DISCUSSION_ATTACHMENT_PATH, null);
                }

                if (attachment != null) {
                    attachments.add(attachment);
                } else {
                    // TODO Show error notification?
                }
                broadCastUploadCompleted(fso);
            }

            // Submit fileIds to the assignment
            if (ACTION_ASSIGNMENT_SUBMISSION.equals(action)) {
                submitFilesForSubmission(courseId, assignmentId, attachments);
            } else if (ACTION_DISCUSSION_ATTACHMENT.equals(action)) {
                broadCastDiscussionSuccess(attachments);
            } else {
                updateNotificationComplete();
                broadCastAllUploadsCompleted(attachments);
            }

        } catch (Exception exception) {
            updateNotification(getString(R.string.error_uploading_file));
            broadcastError(exception.getMessage());
        }

        stopSelf();
    }

    private void submitFilesForSubmission(long courseId, long assignmentId, ArrayList<RemoteFile> attachmentsIds) {
        // TODO
    }

    private RemoteFile uploadSubmissionFiles(long courseId, long assignmentId, long size, String fileName, String contentType, String path) {
        // TODO
        return null;
    }

    private RemoteFile uploadCourseFiles(long courseId, Long parentFolderId, long size, String fileName, String contentType, String path, @Nullable String parentFolderPath) throws IOException, NetworkErrorException {

        if (parentFolderPath != null && parentFolderId != null) throw new IllegalArgumentException("Cannot specify both parentFolderPath and parentFolderId");

        FileUploadParams uploadParams;
        if (parentFolderId != null) {
            uploadParams = FileFolderManager.getFileUploadParamsSynchronous(courseId, fileName, size, contentType, parentFolderId);
        } else {
            uploadParams = FileFolderManager.getFileUploadParamsSynchronous(courseId, fileName, size, contentType, parentFolderPath);
        }

        if (uploadParams == null) throw new NetworkErrorException("Unable to obtain file upload parameters");

        return FileFolderManager.uploadFileSynchronousNoRedirect(uploadParams.getUploadUrl(),
                uploadParams.getPlainTextUploadParams(),
                contentType,
                new File(path));
    }

    private RemoteFile uploadUserFiles(long size, String fileName, String contentType, String path, @Nullable String parentFolderPath, @Nullable Long parentFolderId) throws IOException, NetworkErrorException {

        if (parentFolderPath != null && parentFolderId != null) throw new IllegalArgumentException("Cannot specify both parentFolderPath and parentFolderId");

        FileUploadParams uploadParams;
        if (parentFolderId != null) {
            uploadParams = UserManager.getFileUploadParamsSynchronous(fileName, size, contentType, parentFolderId);
        } else {
            uploadParams = UserManager.getFileUploadParamsSynchronous(fileName, size, contentType, parentFolderPath);
        }

        if (uploadParams == null) throw new NetworkErrorException("Unable to obtain file upload parameters");

        return UserManager.uploadUserFileSynchronous(uploadParams.getUploadUrl(),
                uploadParams.getPlainTextUploadParams(),
                contentType,
                new File(path));
    }

    //endregion

    //region Notifications

    private void broadCastUploadCompleted(FileSubmitObject fso) {
        Bundle bundle = new Bundle();
        bundle.putParcelable(Const.FILENAME, fso);
        final Intent status = new Intent(UPLOAD_COMPLETED);
        status.putExtras(bundle);
        sendBroadcast(status);
    }

    private void broadCastAllUploadsCompleted(ArrayList<RemoteFile> attachments) {
        final Intent status = new Intent(ALL_UPLOADS_COMPLETED);
        status.putParcelableArrayListExtra(Const.ATTACHMENTS, attachments);
        sendBroadcast(status);
    }

    private void broadCastDiscussionSuccess(ArrayList<RemoteFile> attachments) {
        updateNotificationComplete();
        final Intent status = new Intent(ALL_UPLOADS_COMPLETED);
        status.putExtra(Const.ATTACHMENTS, attachments);
        sendBroadcast(status);
    }


    private void broadcastError(String message) {
        updateNotificationError(message);
        Bundle bundle = new Bundle();
        Intent status = new Intent(UPLOAD_ERROR);

        bundle.putString(Const.MESSAGE, message);
        status.putExtras(bundle);
        sendBroadcast(status);
    }

    private void showNotification(int size) {
        mNotificationManager = (NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE);

        mNotificationBuilder = new NotificationCompat.Builder(this)
                .setSmallIcon(R.mipmap.canvas_logo_white)
                .setContentTitle(String.format(Locale.US, getString(R.string.uploading_file_num), 1, size))
                .setProgress(0, 0, true);
        startForeground(NOTIFICATION_ID, mNotificationBuilder.build());
    }

    private void updateNotificationCount(final String fileName, int currentItem) {

        mNotificationBuilder.setContentTitle(String.format(Locale.US, getString(R.string.uploading_file_num), currentItem, mUploadCount))
                .setContentText(fileName);
        mNotificationManager.notify(NOTIFICATION_ID, mNotificationBuilder.build());
    }


    private void updateNotification(final String message) {
        mNotificationBuilder.setContentText(message);
        mNotificationManager.notify(NOTIFICATION_ID, mNotificationBuilder.build());
    }

    private void updateNotificationError(final String message) {
        mNotificationBuilder.setContentText(message)
                .setProgress(0, 0, false);
        mNotificationManager.notify(NOTIFICATION_ID, mNotificationBuilder.build());
    }

    private void updateNotificationComplete() {
        mNotificationBuilder.setProgress(0, 0, false)
                .setContentTitle(getString(R.string.files_uploaded_successfully));
        mNotificationManager.notify(NOTIFICATION_ID, mNotificationBuilder.build());
    }

    private void updateSubmissionComplete() {
        mNotificationBuilder.setProgress(0, 0, false)
                .setContentTitle(getString(R.string.files_submitted_successfully));
        mNotificationManager.notify(NOTIFICATION_ID, mNotificationBuilder.build());
    }

    private void updateMessageComplete() {
        mNotificationBuilder.setProgress(0, 0, false)
                .setContentTitle(getString(R.string.message_sent_successfully));
        mNotificationManager.notify(NOTIFICATION_ID, mNotificationBuilder.build());
    }

    //endregion

    @Override
    public void onDestroy() {
        if(mNotificationManager != null) {
            if (isCanceled) {
                mNotificationManager.cancel(NOTIFICATION_ID);
            } else {
                mNotificationBuilder.setOngoing(false);
                mNotificationManager.notify(NOTIFICATION_ID, mNotificationBuilder.build());
            }
        }
        FileUploadUtils.deleteTempDirectory(getApplicationContext());
    }

    //region Helpers

    public static Bundle getUserFilesBundle(ArrayList<FileSubmitObject> fileSubmitObjects, Long parentFolderID) {
        //Long size, String fileName, String contentType, String path
        Bundle bundle = new Bundle();
        bundle.putParcelableArrayList(Const.FILES, fileSubmitObjects);

        if (parentFolderID != null) {
            bundle.putLong(Const.PARENT_FOLDER_ID, parentFolderID);
        }

        return bundle;
    }

    public static Bundle getCourseFilesBundle(ArrayList<FileSubmitObject> fileSubmitObjects, long courseId, Long parentFolderID) {
        Bundle bundle = new Bundle();
        bundle.putParcelableArrayList(Const.FILES, fileSubmitObjects);
        bundle.putLong(Const.COURSE_ID, courseId);
        if (parentFolderID != null) {
            bundle.putLong(Const.PARENT_FOLDER_ID, parentFolderID);
        }

        return bundle;
    }

    public static Bundle getAssignmentSubmissionBundle(ArrayList<FileSubmitObject> fileSubmitObjects, long courseId, long assignmentId) {
        Bundle bundle = new Bundle();
        bundle.putParcelableArrayList(Const.FILES, fileSubmitObjects);
        bundle.putLong(Const.COURSE_ID, courseId);
        bundle.putLong(Const.ASSIGNMENT_ID, assignmentId);
        return bundle;
    }

    public static Bundle getMessageBundle(ArrayList<FileSubmitObject> fileSubmitObjects, String messageText, long conversationId) {
        //Long size, String fileName, String contentType, String path
        Bundle bundle = new Bundle();
        bundle.putParcelableArrayList(Const.FILES, fileSubmitObjects);
        bundle.putLong(Const.CONVERSATION, conversationId);
        bundle.putString(Const.MESSAGE, messageText);

        return bundle;
    }

    //endregion
}

File: pandautils/src/main/java/com/instructure/pandautils/services/FileUploadService.kt
/*
 * Copyright (C) 2017 - present Instructure, Inc.
 *
 *     Licensed under the Apache License, Version 2.0 (the "License");
 *     you may not use this file except in compliance with the License.
 *     You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 *     Unless required by applicable law or agreed to in writing, software
 *     distributed under the License is distributed on an "AS IS" BASIS,
 *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *     See the License for the specific language governing permissions and
 *     limitations under the License.
 *
 */

package com.instructure.pandautils.services

import android.app.IntentService
import android.app.NotificationChannel
import android.app.NotificationManager
import android.content.Context
import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.os.Parcelable
import android.support.v4.app.NotificationCompat
import com.instructure.canvasapi2.managers.*
import com.instructure.canvasapi2.models.Assignment
import com.instructure.canvasapi2.models.Attachment
import com.instructure.canvasapi2.models.Conversation
import com.instructure.canvasapi2.utils.ContextKeeper
import com.instructure.pandautils.R
import com.instructure.pandautils.models.FileSubmitObject
import com.instructure.pandautils.utils.Const
import com.instructure.pandautils.utils.FileUploadUtils
import java.util.*

class FileUploadService @JvmOverloads constructor(name: String = FileUploadService::class.java.simpleName) : IntentService(name) {

    private var uploadCount: Int = 0

    lateinit private var notificationBuilder: NotificationCompat.Builder
    private val notificationManager by lazy { getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager }

    override fun onHandleIntent(intent: Intent) {
        val action = intent.action
        val bundle = intent.extras
        val fileSubmitObjects = bundle.getParcelableArrayList<FileSubmitObject>(Const.FILES) ?: return
        val assignment = bundle.getParcelable<Assignment>(Const.ASSIGNMENT)

        uploadCount = fileSubmitObjects.size
        showNotification(uploadCount)

        if (assignment != null && assignment.groupCategoryId != 0L) {
            // This is a group assignment, we need to get the list of groups before starting uploads
            GroupManager.getGroupsSynchronous(true)
                    ?.find { it.groupCategoryId == assignment.groupCategoryId }
                    ?.let { startUploads(action, fileSubmitObjects, bundle, it.id) }
                    ?: broadcastError(getString(R.string.errorSubmittingFiles))
        } else {
            startUploads(action, fileSubmitObjects, bundle, null)
        }
    }

    // region Upload
    private fun startUploads(action: String, fileSubmitObjects: ArrayList<FileSubmitObject>, bundle: Bundle, groupId: Long?) {
        val courseId = bundle.getLong(Const.COURSE_ID)
        val assignment = bundle.getParcelable<Assignment>(Const.ASSIGNMENT)
        val conversationId = bundle.getLong(Const.CONVERSATION)
        val quizQuestionId = bundle.getLong(Const.QUIZ_ANSWER_ID, -1)
        val quizId = bundle.getLong(Const.QUIZ)
        val position = bundle.getInt(Const.POSITION)
        val messageText = bundle.getString(Const.MESSAGE)
        val subject = bundle.getString(Const.SUBJECT)
        val isGroup = bundle.getBoolean(Const.IS_GROUP)
        val userIds = bundle.getStringArrayList(Const.USER_IDS)
        val contextId = bundle.getString(Const.CONTEXT_ID)
        val parentFolderId = if (bundle.containsKey(Const.PARENT_FOLDER_ID)) bundle.getLong(Const.PARENT_FOLDER_ID) else null

        val attachmentsIds = mutableListOf<Long>()
        val attachments = mutableListOf<Attachment>()

        try {
            fileSubmitObjects.forEachIndexed { idx, fso ->
                updateNotificationCount(fso.name, idx + 1)
                val config = FileUploadConfig(fso.name, fso.fullPath, fso.size, fso.contentType)
                when (action) {
                    ACTION_ASSIGNMENT_SUBMISSION -> {
                        val uploadContext = if (groupId == null) SubmissionUploadContext(courseId, assignment!!.id) else GroupUploadContext(groupId)
                        attachmentsIds += FileUploadManager.uploadFileSynchronous(uploadContext, config)!!.id
                    }
                    ACTION_COURSE_FILE -> {
                        config.parentFolderId = parentFolderId
                        attachmentsIds += FileUploadManager.uploadFileSynchronous(CourseUploadContext(courseId), config)!!.id
                    }
                    ACTION_USER_FILE -> {
                        config.parentFolderId = parentFolderId
                        attachmentsIds += FileUploadManager.uploadFileSynchronous(UserUploadContext(), config)!!.id
                    }
                    ACTION_MESSAGE_ATTACHMENTS, ACTION_NEW_MESSAGE_ATTACHMENTS -> {
                        config.parentFolderPath = MESSAGE_ATTACHMENT_PATH
                        attachmentsIds += FileUploadManager.uploadFileSynchronous(UserUploadContext(), config)!!.id
                    }
                    ACTION_QUIZ_FILE -> {
                        attachments += FileUploadManager.uploadFileSynchronous(QuizUploadContext(courseId, quizId), config)!!
                    }
                    ACTION_DISCUSSION_ATTACHMENT -> {
                        config.parentFolderPath = DISCUSSION_ATTACHMENT_PATH
                        attachments += FileUploadManager.uploadFileSynchronous(UserUploadContext(), config)!!
                    }
                }
                broadcastUploadCompleted(fso)
            }
            // Submit fileIds to the assignment
            when (action) {
                ACTION_ASSIGNMENT_SUBMISSION -> submitAttachmentsForSubmission(courseId, assignment, attachmentsIds)
                ACTION_MESSAGE_ATTACHMENTS -> submitAttachmentsForMessage(conversationId, messageText, attachmentsIds)
                ACTION_DISCUSSION_ATTACHMENT -> broadcastDiscussionSuccess(attachments)
                ACTION_QUIZ_FILE -> broadcastQuizSuccess(attachments[0], quizQuestionId, position)
                ACTION_NEW_MESSAGE_ATTACHMENTS -> submitAttachmentsForNewMessage(messageText, subject, userIds, contextId, isGroup, attachmentsIds)
                else -> {
                    updateNotificationComplete()
                    broadCastAllUploadsCompleted()
                }
            }
        } catch (exception: Exception) {
            updateNotification(getString(R.string.errorUploadingFile))
            if (quizQuestionId != -1L) {
                broadcastQuizError(exception.message.orEmpty(), quizQuestionId, position)
            } else {
                broadcastError(exception.message.orEmpty())
            }
        }

        stopSelf()
    }

    private fun submitAttachmentsForMessage(conversationID: Long, messageText: String?, attachmentsIds: List<Long>) {
        ConversationManager.addMessageToConversationSynchronous(conversationID, messageText, attachmentsIds)?.let {
            updateMessageComplete()
            broadCastMessageSuccess(it)
        } ?: broadcastError(getString(R.string.errorSendingMessage))
    }

    private fun submitAttachmentsForNewMessage(messageText: String?, subject: String?, userIds: List<String>, contextId: String?, isGroup: Boolean, attachmentsIds: List<Long>) {
        ConversationManager.createConversationWithAttachmentSynchronous(messageText, subject, userIds, contextId, isGroup, attachmentsIds)?.let {
            updateMessageComplete()
            broadCastNewMessageSuccess()
        } ?: broadcastError(getString(R.string.errorSendingMessage))
    }

    private fun submitAttachmentsForSubmission(courseId: Long, assignment: Assignment, attachmentsIds: List<Long>) {
        SubmissionManager.postSubmissionAttachmentsSynchronous(courseId, assignment.id, attachmentsIds)?.let {
            updateSubmissionComplete()
            broadCastAllUploadsCompleted()
        } ?: broadcastError(getString(R.string.errorSubmittingFiles))
    }
    // endregion Upload

    // region Notifications
    private fun broadcastUploadCompleted(fso: FileSubmitObject) {
        val bundle = Bundle()
        bundle.putParcelable(Const.FILENAME, fso)
        val status = Intent(UPLOAD_COMPLETED)
        status.putExtras(bundle)
        sendBroadcast(status)
    }

    private fun broadCastAllUploadsCompleted() {
        val status = Intent(ALL_UPLOADS_COMPLETED)
        sendBroadcast(status)
    }

    private fun broadCastMessageSuccess(conversation: Conversation) {
        val status = Intent(ALL_UPLOADS_COMPLETED)
        status.putExtra(Const.CONVERSATION, conversation as Parcelable)
        sendBroadcast(status)
    }

    private fun broadCastNewMessageSuccess() {
        val status = Intent(ALL_UPLOADS_COMPLETED)
        sendBroadcast(status)
    }

    private fun broadcastDiscussionSuccess(attachments: List<Attachment>) {
        updateNotificationComplete()
        val status = Intent(ALL_UPLOADS_COMPLETED)
        status.putParcelableArrayListExtra(Const.ATTACHMENTS, ArrayList(attachments))
        sendBroadcast(status)
    }

    private fun broadcastQuizSuccess(attachment: Attachment, quizQuestionId: Long, position: Int) {
        updateNotificationComplete()
        val status = Intent(QUIZ_UPLOAD_COMPLETE)
        status.putExtra(Const.ATTACHMENT, attachment as Parcelable)
        status.putExtra(Const.QUIZ_ANSWER_ID, quizQuestionId)
        status.putExtra(Const.POSITION, position)
        sendBroadcast(status)
    }

    private fun broadcastQuizError(message: String, quizQuestionId: Long, position: Int) {
        updateNotificationError(message)
        val bundle = Bundle()
        val status = Intent(UPLOAD_ERROR)
        status.putExtra(Const.QUIZ_ANSWER_ID, quizQuestionId)
        status.putExtra(Const.POSITION, position)
        bundle.putString(Const.MESSAGE, message)
        status.putExtras(bundle)
        sendBroadcast(status)
    }

    private fun broadcastError(message: String) {
        updateNotificationError(message)
        val bundle = Bundle()
        val status = Intent(UPLOAD_ERROR)

        bundle.putString(Const.MESSAGE, message)
        status.putExtras(bundle)
        sendBroadcast(status)
    }

    private fun showNotification(size: Int) {
        createNotificationChannel(CHANNEL_ID)
        notificationBuilder = NotificationCompat.Builder(this, CHANNEL_ID)
                .setSmallIcon(R.drawable.ic_notification_canvas_logo)
                .setContentTitle(String.format(Locale.US, getString(R.string.uploadingFileNum), 1, size))
                .setProgress(0, 0, true)
        startForeground(NOTIFICATION_ID, notificationBuilder.build())
    }

    private fun updateNotificationCount(fileName: String, currentItem: Int) {
        notificationBuilder.setContentTitle(String.format(Locale.US, getString(R.string.uploadingFileNum), currentItem, uploadCount))
                .setContentText(fileName)
        notificationManager.notify(NOTIFICATION_ID, notificationBuilder.build())
    }


    private fun updateNotification(message: String) {
        notificationBuilder.setContentText(message)
        notificationManager.notify(NOTIFICATION_ID, notificationBuilder.build())
    }

    private fun updateNotificationError(message: String) {
        notificationBuilder.setContentText(message)
                .setProgress(0, 0, false)
        notificationManager.notify(NOTIFICATION_ID, notificationBuilder.build())
    }

    private fun updateNotificationComplete() {
        notificationBuilder.setProgress(0, 0, false)
                .setContentTitle(getString(R.string.filesUploadedSuccessfully))
        notificationManager.notify(NOTIFICATION_ID, notificationBuilder.build())
    }

    private fun updateSubmissionComplete() {
        notificationBuilder.setProgress(0, 0, false)
                .setContentTitle(getString(R.string.filesSubmittedSuccessfully))
        notificationManager.notify(NOTIFICATION_ID, notificationBuilder.build())
    }

    private fun updateMessageComplete() {
        notificationBuilder.setProgress(0, 0, false)
                .setContentTitle(getString(R.string.messageSentSuccessfully))
        notificationManager.notify(NOTIFICATION_ID, notificationBuilder.build())
    }

    private fun createNotificationChannel(channelId: String) {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) return

        // Prevents recreation of notification channel if it exists.
        if (notificationManager.notificationChannels.any { it.id == channelId }) return

        val name = ContextKeeper.appContext.getString(R.string.notificationChannelNameFileUploadsName)
        val description = ContextKeeper.appContext.getString(R.string.notificationChannelNameFileUploadsDescription)

        // Create the channel and add the group
        val importance = NotificationManager.IMPORTANCE_HIGH
        val channel = NotificationChannel(channelId, name, importance)
        channel.description = description
        channel.enableLights(false)
        channel.enableVibration(false)

        // Create the channel
        notificationManager.createNotificationChannel(channel)
    }
    // endregion Notifications

    override fun onDestroy() {
        notificationBuilder.setOngoing(false)
        notificationManager.notify(NOTIFICATION_ID, notificationBuilder.build())
        FileUploadUtils.deleteDirectory(FileUploadUtils.getTempFolder(this))
    }

    companion object {
        private const val NOTIFICATION_ID = 1
        const val CHANNEL_ID = "fileUploadChannel"

        // Upload broadcasts
        const val ALL_UPLOADS_COMPLETED = "ALL_UPLOADS_COMPLETED"
        const val QUIZ_UPLOAD_COMPLETE = "QUIZ_UPLOAD_COMPLETE"
        const val UPLOAD_COMPLETED = "UPLOAD_COMPLETED"
        const val UPLOAD_ERROR = "UPLOAD_ERROR"

        // Upload Actions
        const val ACTION_ASSIGNMENT_SUBMISSION = "ACTION_ASSIGNMENT_SUBMISSION"
        const val ACTION_MESSAGE_ATTACHMENTS = "ACTION_MESSAGE_ATTACHMENTS"
        const val ACTION_NEW_MESSAGE_ATTACHMENTS = "ACTION_NEW_MESSAGE_ATTACHMENTS"
        const val ACTION_COURSE_FILE = "ACTION_COURSE_FILE"
        const val ACTION_USER_FILE = "ACTION_USER_FILE"
        const val ACTION_QUIZ_FILE = "ACTION_QUIZ_FILE"
        const val ACTION_DISCUSSION_ATTACHMENT = "ACTION_DISCUSSION_ATTACHMENT"

        private const val MESSAGE_ATTACHMENT_PATH = "conversation attachments"
        private const val DISCUSSION_ATTACHMENT_PATH = "discussion attachments"

        @JvmStatic
        fun getUserFilesBundle(
                fileSubmitObjects: ArrayList<FileSubmitObject>,
                parentFolderId: Long?
        ) = Bundle().apply {
            putParcelableArrayList(Const.FILES, fileSubmitObjects)
            parentFolderId?.let { putLong(Const.PARENT_FOLDER_ID, it) }
        }

        @JvmStatic
        fun getQuizFileBundle(
                fileSubmitObjects: ArrayList<FileSubmitObject>,
                parentFolderId: Long?,
                quizQuestionId: Long,
                position: Int,
                courseId: Long,
                quizId: Long
        ) = Bundle().apply {
            putParcelableArrayList(Const.FILES, fileSubmitObjects)
            putLong(Const.QUIZ_ANSWER_ID, quizQuestionId)
            putLong(Const.QUIZ, quizId)
            putLong(Const.COURSE_ID, courseId)
            putInt(Const.POSITION, position)
            parentFolderId?.let { putLong(Const.PARENT_FOLDER_ID, it) }
        }

        @JvmStatic
        fun getCourseFilesBundle(
                fileSubmitObjects: ArrayList<FileSubmitObject>,
                courseId: Long,
                parentFolderId: Long?
        ) = Bundle().apply {
            putParcelableArrayList(Const.FILES, fileSubmitObjects)
            putLong(Const.COURSE_ID, courseId)
            parentFolderId?.let { putLong(Const.PARENT_FOLDER_ID, it) }
        }

        @JvmStatic
        fun getAssignmentSubmissionBundle(
                fileSubmitObjects: ArrayList<FileSubmitObject>,
                courseId: Long,
                assignment: Assignment
        ) = Bundle().apply {
            putParcelableArrayList(Const.FILES, fileSubmitObjects)
            putLong(Const.COURSE_ID, courseId)
            putParcelable(Const.ASSIGNMENT, assignment)
        }

        @JvmStatic
        fun getMessageBundle(
                fileSubmitObjects: ArrayList<FileSubmitObject>,
                messageText: String,
                conversationId: Long
        ) = Bundle().apply {
            putParcelableArrayList(Const.FILES, fileSubmitObjects)
            putLong(Const.CONVERSATION, conversationId)
            putString(Const.MESSAGE, messageText)
        }

        @JvmStatic
        fun getNewMessageBundle(
                fileSubmitObjects: ArrayList<FileSubmitObject>,
                userIds: ArrayList<String>,
                subject: String,
                messageText: String,
                isGroup: Boolean,
                contextId: String
        ) = Bundle().apply {
            putParcelableArrayList(Const.FILES, fileSubmitObjects)
            putStringArrayList(Const.USER_IDS, userIds)
            putString(Const.SUBJECT, subject)
            putString(Const.MESSAGE, messageText)
            putBoolean(Const.IS_GROUP, isGroup)
            putString(Const.CONTEXT_ID, contextId)
        }
    }
}
